### PR TITLE
support `on update current_timestamp`

### PR DESCRIPTION
**Pull request by JCOR11599**
This PR adds support for handling and executing `ON UPDATE` column expressions, which are triggered when an update is applied to any column in the row.

TODO:
 - we support `now()` for default, but not the syntax for `on update`

fixes https://github.com/dolthub/dolt/issues/6030

### CODE REVIEW

**Review by max-hoffman — COMMENTED** [review]
i think we need a lot more testing, can't just do the one case in the user report. Can we do different field types, INSERTS that trigger ON DUPLICATE UPDATE, update that explicitly sets the ON UPDATE field isn't overwritten by the duplicate, etc

**Review by max-hoffman — APPROVED** [review]
LGTM, I think the extra tests are a good walkthrough of expected behavior for anyone else looking at this feature, and the implementation falls out pretty directly from the behavior needed to support the tests. We'll need Dolt side changes for CREATE TABLE and storage change, and i'd do the Dolt side before merging any of this because of the storage format version issues.

Seems like a minor release bump, new data written with this feature will fail to update using old client code.

File: sql/planbuilder/dml.go [hunk 1]
```diff
@@ -269,6 +269,25 @@ func (b *Builder) assignmentExprsToExpressions(inScope *scope, e ast.AssignmentE
 				generated := b.resolveColumnDefaultExpression(inScope, col, col.Generated)
 				updateExprs = append(updateExprs, expression.NewSetField(colName, assignColumnIndexes(generated, tableSch)))
 			}
+			if col.OnUpdate != nil {
+				// don't add if column is already being updated
+				colName := expression.NewGetFieldWithTable(i, int(tabId), col.Type, col.DatabaseSource, col.Source, col.Name, col.Nullable)
```
Comment (zachmu) [inline sql/planbuilder/dml.go:274]: Should extract an isColumnUpdated method here

File: sql/planbuilder/dml.go [hunk 2]
```diff
@@ -416,6 +435,8 @@ func (b *Builder) buildUpdate(inScope *scope, u *ast.Update) (outScope *scope) {
 	ignore := u.Ignore != ""
 	update := plan.NewUpdate(outScope.node, ignore, updateExprs)
 
+	outScope.node.Schema()
```
Comment (zachmu) [inline sql/planbuilder/dml.go:438]: ?

**Review by zachmu — APPROVED** [review]
LGTM but see comments

File: enginetest/queries/script_queries.go [hunk 3]
```diff
@@ -6244,3 +6244,463 @@ var DropDatabaseScripts = []ScriptTest{
 		},
 	},
 }
+
+var ZeroTime = time.Date(0000, time.January, 1, 0, 0, 0, 0, time.UTC)
+var SetupTime = time.Date(2000, time.January, 1, 12, 0, 0, 0, time.UTC)
+var QueryTime = time.Date(2023, time.December, 15, 1, 30, 0, 0, time.UTC)
+var OtherTime = time.Date(2020, time.October, 2, 0, 0, 0, 0, time.UTC)
+var OnUpdateExprScripts = []ScriptTest{
```
Comment (zachmu) [inline enginetest/queries/script_queries.go:6252]: Probably put these in update_queries.go instead

File: enginetest/queries/script_queries.go [hunk 4]
```diff
@@ -6244,3 +6244,463 @@ var DropDatabaseScripts = []ScriptTest{
 		},
 	},
 }
+
+var ZeroTime = time.Date(0000, time.January, 1, 0, 0, 0, 0, time.UTC)
+var SetupTime = time.Date(2000, time.January, 1, 12, 0, 0, 0, time.UTC)
```
Comment (zachmu) [inline enginetest/queries/script_queries.go:6249]: I would name these after the time they describe, e.g. Jan1Noon etc.

Comment (zachmu) [reply enginetest/queries/script_queries.go:6249]: Could also make them const

**Review by zachmu — APPROVED** [review]
LGTM, thanks for the changes

File: enginetest/queries/script_queries.go (hunk 4 above)
Comment (JCOR11599) [reply enginetest/queries/script_queries.go]: Renamed, but I golang does not like `time.Date()` as a const

**Pull request merged**